EIPS/eip-xxxx.md

```diff
@@ -0,0 +1,89 @@
+
+These storage usage patterns are invisible to the Solidity compiler because they are not represented as Solidity state variables. Smart contract tools like static analyzers or blockchain explorers often need to know the storage location of contract data. Standardizing the location for non-linear storage layouts will allow these tools to correctly interpret contracts where these design patterns are used.
+
+## Specification
+
+A _namespace_ consists of a set of variables that are placed contiguous in the storage layout of a contract. It should be implemented as a struct.
+
+A _namespace id_ is a string that uniquely identifies a namespace in a contract. It should not contain any whitespace characters.
+
+The storage location for a namespace is defined as `ns_loc(id: string) = keccak256(uint256(keccak256(id)) - 1)`.
+
+A Solidity contract using namespaced storage can annotate a struct with the NatSpec tag `@custom:storage-location erc7201:<NAMESPACE_ID>` to identify it as a namespace with id `<NAMESPACE_ID>`.
+
+## Rationale
+
+A requirement for the location is that it shouldn't overlap with any storage location that would be used by the standard Solidity layout. This is in case namespaced storage is used alongside standard linear storage, either deliberately or accidentally.
+
+First, note that a namespace may be larger than a single storage slot, so a variable in a namespace will be placed in a slot `ns_loc(id) + k`. If we assume collision resistance of Keccak-256, the chosen `ns_loc` function has the desired property with very high probability, because the cases in which a Solidity variable receives a location like it, that is of the form `keccak256(X)`, are:
+
+1. Arrays:
+    1. If the array is at the top level and is variable number `N` in the layout, the location of the `k`th item in the array will be `keccak256(N) + k`. If `N` will be a number much smaller than `uint256(keccak256(id)) - 1`.
+    2. If the array is within another array or mapping, it will be in some location `keccak256(X) + j`, and the `k`th item will be at `keccak256(keccak256(X) + j) + k`. For this to equal `ns_loc(X) + k` we would need `j = -1`, but `j` will always be a positive number in standard Solidity layout.
+2. Mappings: The value for key `q` in a mapping will be at location `keccak256(H(q) .. X)` where `X` is the location of the mapping itself. Note that `H(q)` can be any number of bytes. If it is a non-zero number of bytes, it is distinct from any `ns_loc(id) + k`. If it is zero bytes, it can be that `ns_loc(id) = keccak256(X)` if `X = keccak256(id) - 1`. But we know that `X` is the location of the mapping and as mentioned for arrays above a variable can be at `keccak256(Y) + j` for a positive number `j`.
```

Comment:
```suggestion
2. Mappings: The value for key `q` in a mapping will be at location `keccak256(H(q) .. X)` where `X` is the location of the mapping itself. Note that `H(q)` can be any number of bytes. If it is a non-zero number of bytes, it is distinct from any `ns_loc(id) + k`. If it is zero bytes, it can be that `ns_loc(id) = keccak256(X)` if `X = keccak256(id) - 1`, but we know that `X` is the location of the mapping and (as mentioned for arrays above) a variable will be at `keccak256(Y) + j` for a positive number `j`.
```